### DISCUSSION TITLE
Ransack - ActionView::Template::Error

### DISCUSSION THREAD

**Discussion by o-rashad** [activeadmin/activeadmin#8033] [Help] [answered]
### Did you find a bug?

I'm getting the following error for all my app models in active admin dashboard, only the `Dashboard` tap is working

This is only happening in my production environment in heroku, the local is working fine. 
```
ActionView::Template::Error (Ransack needs AdminUser attributes explicitly allowlisted as searchable. Define a `ransackable_attributes` class method in your `AdminUser` model, watching out for items you DON'T want searchable (for example, `encrypted_password`, `password_reset_token`, `owner` or other sensitive information).
```
### Expected behavior

Show the index table of any of my active admin configured models.

### Actual behavior

Blank page and specified error log in production.

**Comment by pjpires** [comment, marked as answer]
I think the issue comes from the first breaking change listed in ransack's v4 changelog [here](https://github.com/activerecord-hackery/ransack/releases/tag/v4.0.0).

It seems to require your models to declare both `self.ransackable_attributes` and `self.ransackable_associations` methods. This will do the trick:
```ruby
class ApplicationRecord < ActiveRecord::Base
  # ...

  def self.ransackable_attributes(auth_object = nil)
    authorizable_ransackable_attributes
  end

  def self.ransackable_associations(auth_object = nil)
    authorizable_ransackable_associations
  end
end
```
but as far as I understood it means you're whitelisting all your models attributes for searching against. If you want to customize it per model you probably need/should declare in each of your models and return only the attributes you want to whitelist.

**Reply by StanBright** [reply]
Thanks @pjpires. That did the job for me. If it wasn't for your answer, I could have spent hours bumping my head. Cheers.

**Reply by o-rashad** [reply]
Thank you, @pjpires. that solved my issue. 💯 

I really want to know, how did you solve it, for example what was running in your mind when you saw the error, how did you break it down ? how did you track it all way up to the ransack change log !?

I'm just asking to learn from your thinking model for solving first time issues like this.

But anyway I'm thankful for you. 😸

**Reply by tagliala** [reply]
Hello, sorry for stepping in.

It is weird that this only happens in production, unless the development server has not been restarted after a bundle update. Or, `Gemfile.lock` has not been commited and Heroku is downloading a fresh version of Active Admin (compared to the one in development)

However, since Active Admin, like many gems, follows SemVer and just released a new major version 3.x, one of the most important things to check is the changelog. When a new major version is released, then there is probably a breaking change

This change related to Ransack is clearly documented here, in Active Admin's changelog: https://github.com/activeadmin/activeadmin/blob/master/CHANGELOG.md#300-

This particular change has been made for Security (Information Disclosure) issues

**Reply by o-rashad** [reply]
@tagliala, Thats true.

Though, I restarted the local server several times, the error showed up, after a while. I don't know exactly what made the error show up, I tried different things back then, like `bundle update`.

**Comment by pjpires** [comment]
@od-c0d3r my thinking model was aligned to what @tagliala mentioned: when I saw the changelog for 3.0.0, tbh I didn't immediately pay too much attention to it requiring a new ransack version. When it got broken I checked again a bit more thoroughly, then went to ransack's repo and checked the changelog for v4, which mentions "Require explict allowlisting of attributes and associations" as one of the breaking changes, that kinda did it for me. 😄 

@tagliala In my case it was happening in development, too. Had a few gem bumps along the way, so I got the error right after updating to 3.0.0, restarting the server and reloading the page in the admin panel. 👍  BTW, would it be possible somehow/make sense for Active Admin to declare these methods (`ransackable_attributes` and `ransackable_associations`) based on the attributes in use for each of the models? I acknowledge I didn't understand the change to its full extent, so in my mind it made sense, but not sure if that is possible or what implications it could have.

**Comment by cjilbert504** [comment]
This discussion helped me in fixing our issue with the upgrade, so a big thanks to everyone here! 

Extending `ApplicationRecord` for the most part worked, however, in our admin panel we also have some models from gems/engines so those needed to have the methods mentioned in the solution above added to them. It turned out that these actually inherit from `ActiveRecord::Base` versus `ApplicationRecord` so I had to go up a level.

In our case, allowing all attributes/associations to be searchable is ok. With that said, what I ended up doing was creating a module containing the two methods and then extending the module on `ActiveRecord::Base` via an `ActiveSupport::on_load` call in an initializer. This covered all our models plus the gems/engines and kept it all in one place. I wrote up a quick [blog post](https://gorails.com/blog/patching-models-for-ransack-4-0-0-extending-activerecord-for-gem-compatibility) about it in hopes that maybe it will help someone else who needs to do the same. Here's the code though just in case no one wants to read the blog but still see the code:
```ruby
# config/initializers/ransack_patch.rb

module RansackPatch
  def ransackable_attributes(auth_object = nil)
    authorizable_ransackable_attributes
  end

  def ransackable_associations(auth_object = nil)
    authorizable_ransackable_associations
  end
end

ActiveSupport.on_load(:active_record) do
  extend RansackPatch
end
```

**Reply by viktorsmari** [reply]
Nice one @cjilbert504 
How would you ban searching for `password` on the `User` model?
Would that work with this module?

In our app we have 50 different models (which `belong_to` a `User`) all with a bunch of variables and everything is searchable **except** the `User` model.

**Reply by cjilbert504** [reply]
@viktorsmari - Thanks for the kind words!

 I think you have two options for not allowing the `password` attribute to be searchable. 

If you want to keep everything within the module I shared I think you could change the `ransackable_attributes`method in the module to do something like this: 
```ruby
# config/initializers/ransack_patch.rb

module RansackPatch
  def ransackable_attributes(auth_object = nil)
    authorizable_ransackable_attributes - ["password"]
  end

  def ransackable_associations(auth_object = nil)
    authorizable_ransackable_associations
  end
end

ActiveSupport.on_load(:active_record) do
  extend RansackPatch
end
```

This approach would remove the `password` attribute from ALL models in the allowed list of searchable attributes. If a model doesn't have a `password` attribute, this won't raise any error, it simple will return all the attributes.

If you want to keep it scoped to the User model, I think you would just need to override the `ransackable_attributes` method in the `User` model like above, for example:
```ruby
class User < ApplicationRecord
  def self.ransackable_attributes(auth_object = nil)
    super - ["password"]
  end
end
```
